### DISCUSSION TITLE
Welcome to fiction Discussions!

### DISCUSSION THREAD

**Discussion by marcelwa** [cda-tum/fiction#8] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by nijianguo2021** [comment]
Dear Professor,
Excuse me, I'm using your software fiction at the moment AND I want to ask you a few questions. First of all, fiction is used in FCN circuit physical design AND only MAJ, OR,AND NOT gates are available. Could you add some new ones?Secondly, whether there is room for improvement in Exact method based on SMT and OGD, and whether a new clock scheme can be added in the end.
Or what else can I do? Please let me know. Thank you！

**Reply by marcelwa** [reply]
Hello and thank you for your message! I'm not a professor, you can call me Marcel. Feel free to fork fiction and implement your changes and improvements.

New clocking schemes can easily be added in `src/topo/fcn_clocking_scheme.h`. If you want to use them in physical design algorithms like `exact`, add them to the look-up function `get_clocking_scheme` in `src/topo/fcn_clocking_scheme.cpp`.

Currently, there is no plan to add further gate functions as long as there are no physically proven gate libraries that support said elements. However, this might change when support for the SiDB technology is added with a respective output format for [SiQAD](https://github.com/siqad/siqad).

I bet there is plenty of room for improvement in both `exact` and `ortho`. However, `exact` requires a deep understanding of the SMT encoding. The `ortho` algorithm, on the other hand, might be the easier start. You can find details on how it works in [the paper](http://www.informatik.uni-bremen.de/agra/doc/konf/2019_ASP-DAC_Scalable_Design_for_Field-coupled_Nanocomputing_Circuits.pdf).

Also, you can implement your own physical design algorithms by inheriting from the `physical_design` class in `src/algo/physical_design.h`, adding your command for the CLI in the `src/io/cmd` folder, and adding it to the include list in `src/io/commands.h`. See, e.g., `src/io/cmd/exact.h` for inspiration.

Feel free to reach out again if you have further questions.

**Reply by marcelwa** [reply]
With the recent release of v0.4.0, this comment is now outdated. Quick update:

- clocking schemes can be found in `include/fiction/layouts/clocking_scheme.hpp` and so can the `get_clocking_scheme` look-up function
- the SiDB technology is now fully supported together with a file format writer for SiQAD and a first gate library 
- there is no need for custom physical design algorithms to inherit from a certain class any longer
- the CLI commands are now to be found at `cli/cmd/`

**Comment by nijianguo2021** [comment]
Dear Marcel ,
Thank you for your guidance.

**Comment by leonardoaramburu** [comment]
Dear professor Marcel,

I'm having trouble compiling the fiction tool. When I run the cmake command CMakeList.txt, the fiction tool gives me this error:
"CMake Error: The source directory "/home/leonardo/fiction/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI."
Can you give me some guidance?

Thank you!

ps.: Sorry for my english.

**Reply by marcelwa** [reply]
Dear Leonardo,

Thanks for your interest in the fiction framework and thank you for your message. Since I'm not a professor, you can just call me Marcel. 

From the error message you have posted I get that you have followed the compilation process outlined in the documentation. It appears that the CMake call is failing. Please be sure to execute CMake in the build directory as follows:
```
cmake ..
```
Notice the two dots `..` that refer to the parent directory where there CMakeLists.txt file is located. 

Please let me know if calling CMake with those two dots as argument still gives you the same issue.

Good luck and best regards!

**Comment by leonardoaramburu** [comment]
Dear Marcel,
Good afternoon!

I ran the CMake ../CMakeLists.txt command inside the build directory. The system returned an error, which can be seen in the attachment.

![Captura de tela de 2022-03-03 15-19-43](https://user-images.githubusercontent.com/62958230/156627373-c2171bff-867c-457d-a61a-b06cdf0f533a.png)

I tried running cmake .. command, and it returned the error which can be seen in the attachment.

![Captura de tela de 2022-03-03 15-36-22](https://user-images.githubusercontent.com/62958230/156630050-98142bab-780a-420f-9ffd-d6ba628feaf4.png)

Thanks for your attention and help!

Good luck and best regards!

**Reply by marcelwa** [reply]
Thank you for the clarifications and the screenshots. They were very helpful in determining possible sources of the errors you're facing.

The `CMakeLists.txt` file should not be targeted by the `cmake` command directly. Instead, the folder that `CMakeLists.txt` is located in should be passed as an argument (`..`). Since you have first called `cmake ../CMakeLists.txt`, you might have corrupted your build directory such that the consecutive (otherwise correct) call to `cmake ..` did not work properly either.

Please attempt the following:
- Navigate to your build directory: `cd fiction/build/`
- Remove all files within it: `rm -rf *`
- Retrigger the build process via the correct CMake call: `cmake ..`
- Follow the next steps of the build instructions, e.g., `make`

Please don't hesitate to reach back in case you encounter further isseus. Good luck!

**Reply by leonardoaramburu** [reply]
Dear Marcel. 
I will describe the commands I am doing to run a dumb tool.

1. git clone --recursive https://github.com/marcelwa/fiction.git
2. cd fiction
3. build mkdir
4. cmake ../CMakeLists.txt

But, the command returns the following message:
![Captura de tela de 2022-03-04 18-32-48](https://user-images.githubusercontent.com/62958230/156845221-bda97c4e-d2a3-489f-aa6c-09df99bef641.png)

Sorry for the inconvenience.

Regards

**Reply by marcelwa** [reply]
Dear Leonardo,

The commands should instead be
```
git clone --recursive https://github.com/marcelwa/fiction.git
cd fiction
mkdir build
cmake ..
```
in a fresh folder. Please note to **not** write `cmake ../CMakeLists.txt`

**Comment by leonardoaramburu** [comment]
Dear Marcel,

It's working. 

Thank you!

Regards.

**Comment by leonardoaramburu** [comment]
Dear Marcel,
I am unable to run the EXACT in fiction> command. Below you can see the return.

![Captura de tela de 2022-03-21 16-09-24](https://user-images.githubusercontent.com/62958230/159346480-c12ae8c7-8ca1-46d1-bedf-bd243331616e.png)
Regards.

**Reply by marcelwa** [reply]
Hi Leonardo,
Please make sure that you have installed the Z3 solver and compiled *fiction* with the respective CMake flag. You can find further information [here](https://fiction.readthedocs.io/en/latest/getting_started.html#smt-based-exact-p-r).

Good luck!
